### PR TITLE
gh-852: Update SampleIntegrationTests to run tests against packaged CLI

### LINKED ISSUE #852: Update SampleIntegrationTests to run tests against packaged CLI
`SampleIntegrationTests` currently runs the samples using the classpath in the IDE or in Maven. This classpath is subtly different to the classpath that's available when using the packaged CLI. This can cause problems. For example template.groovy currently fails when run with the CLI, even though the test that runs it currently passes. Updating the tests to use `CommandLineInvoker` will run them against the packaged CLI and flush out any such classpath issues.

### PR DESCRIPTION
**Pull request by tkopczynski**
This commit updates integration tests for Spring Boot CLI samples so that they use `CommandLineInvoker`.

One thing that might not be obvious is the handling of shutting down the server. As far as I know it is not possible to kill a process and all of its child processes with Java 8 API consistently on all OSes. Unfortunately, this is the case in these integration tests because `CommandLineInvoker` calls the `spring` command which then runs a `java` child process. The only handle to a process that we're having is the `spring` command process handle. So, when `java` process is a server process and stays there waiting for incoming connections, then it stays in the background even after killing the base `spring` command process.
The solution that I've implemented is to add an additional `shutdown` endpoint to all those server processes which gets called after each test. It shuts down the context and process exits as required.

Hopefully, this is acceptable.

Fixes #852

### CODE REVIEW

**Comment by tkopczynski** [comment]
CI errors may seem related but I didn't add StringUtils usage anywhere, so maybe there is an option to run CI build again?

**Comment by wilkinsona** [comment]
@tkopczynski Thanks for the PR. Unfortunately, Travis is rather unreliable. It regularly corrupts jars when they are downloaded. That, I believe, is what's causing the failure here. I'll close and then re-open the PR to trigger another build.

**Comment by tkopczynski** [comment]
Thanks @wilkinsona it looks better now.

File: spring-boot-cli/src/it/java/org/springframework/boot/cli/infrastructure/CommandLineInvoker.java [hunk 1]
```diff
@@ -187,6 +193,35 @@ public int await() throws InterruptedException {
...
+			}
+		}
+
+		public boolean isAlive() {
+			return this.process.isAlive();
+		}
+
+		public boolean waitForStarted() {
+			Instant started = Instant.now();
+			for (; ;) {
```
Comment (snicoll) [inline spring-boot-cli/src/it/java/org/springframework/boot/cli/infrastructure/CommandLineInvoker.java]: `while(ChronoUnit.MILLIS.between(started, Instant.now()) < TIMEOUT)` looks more sensible to me.

File: spring-boot-cli/samples/template.groovy [hunk 2]
```diff
@@ -2,6 +2,8 @@ package org.test
 
 import static org.springframework.boot.groovy.GroovyTemplate.*;
 
+@Grab("groovy-templates")
+
```
Comment (snicoll) [inline spring-boot-cli/samples/template.groovy]: Please remove this line

File: spring-boot-cli/pom.xml [hunk 3]
```diff
@@ -153,6 +153,11 @@
 			<artifactId>junit</artifactId>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.security</groupId>
+			<artifactId>spring-security-config</artifactId>
+			<scope>test</scope>
+		</dependency>
```
Comment (snicoll) [inline spring-boot-cli/pom.xml]: Why is this dependency necessary now?

File: spring-boot-cli/src/test/java/org/springframework/boot/cli/SampleIntegrationTests.java [hunk 4]
```diff
@@ -16,14 +16,20 @@
...
 import java.net.URI;
 
+import org.hamcrest.Matchers;
+import org.junit.After;
+import org.junit.Assert;
 import org.junit.Ignore;
-import org.junit.Rule;
 import org.junit.Test;
 
-import static org.assertj.core.api.Assertions.assertThat;
```
Comment (snicoll) [inline spring-boot-cli/src/test/java/org/springframework/boot/cli/SampleIntegrationTests.java]: There's really no reason to remove that.

File: spring-boot-cli/src/test/java/org/springframework/boot/cli/SampleIntegrationTests.java [hunk 5]
```diff
@@ -32,112 +38,204 @@
  * @author Greg Turnquist
  * @author Roy Clarkson
  * @author Phillip Webb
+ * @author Tomek Kopczynski
  */
 public class SampleIntegrationTests {
 
-	@Rule
-	public CliTester cli = new CliTester("samples/");
+	private static final int PORT = 9000;
```
Comment (snicoll) [inline spring-boot-cli/src/test/java/org/springframework/boot/cli/SampleIntegrationTests.java]: We can't assume that 9000 will be always free.

File: spring-boot-cli/src/test/java/org/springframework/boot/cli/SampleIntegrationTests.java [hunk 6]
```diff
@@ -32,112 +38,204 @@
...
 	@Test
 	public void appSample() throws Exception {
-		String output = this.cli.run("app.groovy");
+		this.invocation = this.invoker.invoke("run", "samples/app.groovy");
 		URI scriptUri = new File("samples/app.groovy").toURI();
-		assertThat(output).contains("Hello World! From " + scriptUri);
+
+		Assert.assertThat(this.invocation.await(), Matchers.equalTo(0));
+		Assert.assertThat(this.invocation.getErrorOutput().length(), Matchers.equalTo(0));
+		Assert.assertThat(this.invocation.getOutput(), Matchers.containsString("Hello World! From " + scriptUri));
```
Comment (snicoll) [inline spring-boot-cli/src/test/java/org/springframework/boot/cli/SampleIntegrationTests.java]: This is applicable to all the tests that were changed. This made the body of the significantly more complex. Can you please remove those fully qualified call for a start and abstract the await part so that it doesn't have to ve repeated in all tests?

File: spring-boot-cli/src/test/resources/securityConfiguration.groovy [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+@Configuration
+class SecurityConfiguration extends WebSecurityConfigurerAdapter {
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http.authorizeRequests()
+                .antMatchers("/shutdown").permitAll()
+    }
+}
```
Comment (snicoll) [inline spring-boot-cli/src/test/resources/securityConfiguration.groovy]: I am not sure I understand why this change requires some security stuff now.

Comment (tkopczynski) [inline spring-boot-cli/src/test/resources/securityConfiguration.groovy]: It's because of the oauth2 sample. It enables authorization server and all endpoints become secured. As part of this PR I added the shutdown endpoint which lets me shut the server down gracefully. In the case of the oauth2 sample, I just need the same endpoint but without any security constraints.

File: spring-boot-cli/samples/template.groovy (hunk 2 above)
Comment (tkopczynski) [inline spring-boot-cli/samples/template.groovy]: Unfortunately, without this I'm getting an error when running this template:
Caused by: java.lang.NoClassDefFoundError: Unable to load class org.springframework.boot.groovy.GroovyTemplate due to missing dependency groovy/text/TemplateEngine

Adding this line seemed like the only way to make it run

**Comment by tkopczynski** [comment]
@snicoll thanks for the comments. I applied most of them and commented the rest.

**Comment by wilkinsona** [comment]
Thanks for the PR. I'm concerned about the multiple problem and the proposed solution. My fear is that it will destabilise the build. I don't think the benefit of testing about the packaged CLI is worth that risk. Thanks anyway.
